src/net.cpp

```diff
@@ -2019,8 +2019,9 @@ void CConnman::ThreadOpenConnections(const std::vector<std::string> connect)
             // from advertising themselves as a service on another host and
             // port, causing a DoS attack as nodes around the network attempt
             // to connect to it fruitlessly.
-            if (addr.GetPort() != Params().GetDefaultPort() && nTries < 50)
+            if (!addr.IsI2P() && addr.GetPort() != Params().GetDefaultPort() && nTries < 50) {
```

Comment:
I don't think this is exactly what we want. As (with SAM 3.1) we're unable to connect to anything except port 0 (from the perspective of later versions), shouldn't this equally bias away from port 0, as we bias away from the default port for other networks?

In other words: maybe this can be done a bit cleaner by having a network-dependent default port, which is 0 for I2P and the normal port for everything else?

Reply:
I think the limitation of only being able to connect to ports that equal 0 is purely internal to the I2P code - it depends on the SAM version being used and higher level code shouldn't care about that or be aware of it. So I planted the refuse-to-connect-to-non-zero-ports [check in i2p.cpp](https://github.com/bitcoin/bitcoin/pull/22112/commits/4401bb56cf3822950b221c3fa47fb0b65312d12f#diff-8945a03581b372d674f34dcb6a63e537343e376ebcd9d8c7c3d94d4366cd9b9dR175-R180).

This allows some day to remove that by only modifying `i2p.cpp` - easier, compared to if this knowledge is spilled in other modules.

A network-dependent default port seems like an unnecessary generalization to me because it is only relevant for I2P. If there was at least one other network that used !=8333 default port, then it would be more appealing :)

Off-scope: I think this entire `if` should be removed.

Reply:
It may be good to update the comment preceding this line to explain the exception for I2P.

Reply:
Added a comment.

Reply:
From testing and offline discussion with @vasild, it seems a network-dependent `GetDefaultPort()` is indeed needed.

Reply:
So, network-dependent default port seems to be the most elegant solution to [resolve addresses without ports](https://github.com/bitcoin/bitcoin/pull/22112#issuecomment-872224803) properly - e.g. `1.2.3.4` should be treated as `1.2.3.4:8333` and `foo.b32.i2p` as `foo.b32.i2p:0`.

Now this line looks better?


```cpp
if (addr.GetPort() != Params().GetDefaultPort(addr.GetNetwork()) && nTries < 50) {
```